File: 19.03.2023/py/main.py
# Function should convert passing in RGB decimal values in a hexadecimal representation


class WrongType(Exception):
    pass

# Solution one
def rgb_one(r, g, b):
    # Round values that are out of range
    r = max(0, min(255, round(r)))
    g = max(0, min(255, round(g)))
    b = max(0, min(255, round(b)))

    # Convert to hexadecimal string and pad with zeros if necessary
    return '{:02X}{:02X}{:02X}'.format(r, g, b)

# Solution two
def rgb_two(r, g, b):
    # Round values that are out of range
    r = max(0, min(255, round(r)))
    g = max(0, min(255, round(g)))
    b = max(0, min(255, round(b)))

    # Convert to hexadecimal using bit-shifting
    hex_value = (r << 16) + (g << 8) + b

    # Format as a 6-digit hexadecimal string
    return '{:06X}'.format(hex_value)

# Solution three
def rgb_three(r, g, b):
    # Round values that are out of range
    r = max(0, min(255, round(r)))
    g = max(0, min(255, round(g)))
    b = max(0, min(255, round(b)))

    # Convert to hexadecimal using the built-in hex() function
    hex_r = hex(r)[2:].zfill(2)
    hex_g = hex(g)[2:].zfill(2)
    hex_b = hex(b)[2:].zfill(2)

    # Concatenate the hexadecimal values
    hex_value = hex_r + hex_g + hex_b

    return hex_value.upper()

# Solution four
import struct

def rgb_four(r, g, b):
    # Round values that are out of range
    r = max(0, min(255, round(r)))
    g = max(0, min(255, round(g)))
    b = max(0, min(255, round(b)))

    # Pack RGB values into a binary string
    binary_str = struct.pack('BBB', r, g, b)

    # Convert binary string to a hexadecimal string
    hex_str = binary_str.hex()

    return hex_str.upper()

# Solution five
def rgb_five(r, g, b):
    # Round values that are out of range
    r = max(0, min(255, round(r)))
    g = max(0, min(255, round(g)))
    b = max(0, min(255, round(b)))

    # Convert RGB values to hexadecimal digits
    hex_r = hex((r >> 4) * 16 + (r & 15))[2:].zfill(2)
    hex_g = hex((g >> 4) * 16 + (g & 15))[2:].zfill(2)
    hex_b = hex((b >> 4) * 16 + (b & 15))[2:].zfill(2)

    # Concatenate the hexadecimal values
    hex_value = hex_r + hex_g + hex_b

    return hex_value.upper()



import cProfile
import time
from timeit import default_timer as timer


if __name__ == '__main__':
    cProfile.run('cut_log(230, 4, xs)')

    start_time = time.time()
    cut_log(230, 4, xs)
    end_time = time.time()

    print('', end_time - start_time)

File: 19.03.2023/js/main.js
// Function should convert passing in RGB decimal values in a hexadecimal representation


/* Solution one */
export const rgbOne = (r, g, b) => {
    // Round the input values to the nearest valid range [0, 255]
    r = Math.max(0, Math.min(255, Math.round(r)));
    g = Math.max(0, Math.min(255, Math.round(g)));
    b = Math.max(0, Math.min(255, Math.round(b)));

    // Convert the decimal values to their hexadecimal representation
    const hexR = r.toString(16).padStart(2, '0');
    const hexG = g.toString(16).padStart(2, '0');
    const hexB = b.toString(16).padStart(2, '0');

    // Concatenate the three hexadecimal values to form the final output
    return String(hexR + hexG + hexB).toUpperCase();
}

/* Solution two */
export const rgbTwo = (r, g, b) => {
    // Round the input values to the nearest valid range [0, 255]
    r = Math.max(0, Math.min(255, Math.round(r)));
    g = Math.max(0, Math.min(255, Math.round(g)));
    b = Math.max(0, Math.min(255, Math.round(b)));

    // Convert the decimal values to their hexadecimal representation
    const hexR = r.toString(16);
    const hexG = g.toString(16);
    const hexB = b.toString(16);

    // Pad each hexadecimal value with a leading zero if necessary
    const paddedR = hexR.length < 2 ? '0' + hexR : hexR;
    const paddedG = hexG.length < 2 ? '0' + hexG : hexG;
    const paddedB = hexB.length < 2 ? '0' + hexB : hexB;

    // Concatenate the three hexadecimal values to form the final output
    return String(paddedR + paddedG + paddedB).toUpperCase();
}

/* Solution three */
export const rgbThree = (r, g, b) => {
    // Round the input values to the nearest valid range [0, 255]
    r = Math.max(0, Math.min(255, Math.round(r)));
    g = Math.max(0, Math.min(255, Math.round(g)));
    b = Math.max(0, Math.min(255, Math.round(b)));

    // Convert the decimal values to their hexadecimal representation
    const hex = (r << 16) | (g << 8) | b;

    // Convert the hexadecimal value to a string and pad it with leading zeros if necessary
    return String(('000000' + hex.toString(16)).slice(-6)).toUpperCase();
}

/* Solution four */
export const rgbFour = (r, g, b) => {
    // Round the input values to the nearest valid range [0, 255]
    r = Math.max(0, Math.min(255, Math.round(r)));
    g = Math.max(0, Math.min(255, Math.round(g)));
    b = Math.max(0, Math.min(255, Math.round(b)));

    // Define a lookup table for hexadecimal digits
    const hexTable = '0123456789abcdef';

    // Convert each decimal value to its corresponding hexadecimal digit
    const hexR = hexTable[Math.floor(r / 16)] + hexTable[r % 16];
    const hexG = hexTable[Math.floor(g / 16)] + hexTable[g % 16];
    const hexB = hexTable[Math.floor(b / 16)] + hexTable[b % 16];

    // Concatenate the three hexadecimal values to form the final output
    return String(hexR + hexG + hexB).toUpperCase();
}

/* Solution five */
export const rgbFive = (r, g, b) => {
    // Round the input values to the nearest valid range [0, 255]
    r = Math.max(0, Math.min(255, Math.round(r)));
    g = Math.max(0, Math.min(255, Math.round(g)));
    b = Math.max(0, Math.min(255, Math.round(b)));

    // Convert the decimal values to their hexadecimal representation
    const hex = [r, g, b].map(x => x.toString(16).padStart(2, '0')).join('');

    // Use a regular expression to format the output as a six-digit hexadecimal number
    return String(hex.replace(/^([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})$/, '$1$2$3')).toUpperCase();
}
